### PR TITLE
Add [v98] Initial wallpaper feature skeleton

### PR DESCRIPTION
**Pull request by electricRGB**

### CODE REVIEW

**Comment by electricRGB** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

_1) What questions will you answer with this data?_

It answers how our users will be using wallpapers.

_2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?_

This provides information about user behavior on wallpaper use.

_3) What alternative methods did you consider to answer these questions? Why were they not sufficient?_

Current methods have not previously recorded this type of telemetry.

_4) Can current instrumentation answer these questions?_

No.

_5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki._

_**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**_

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Whether or not users tapped the cycle wallpaper button on the homepage.</td>
    <td>Category 2</td>
    <td>#9930</td>
  </tr>
 <tr>
    <td>Whether users selected a custom wallpaper from the settings screen.</td>
    <td>Category 2</td>
    <td>#9930</td>
  </tr>
 <tr>
    <td>Whether users toggled wallpaper homepage togge on/off.</td>
    <td>Category 2</td>
    <td>#9930</td>
  </tr>
</table>

_6) How long will this data be collected?  Choose one of the following:_

Until 2022-07-01.

_7) What populations will you measure?_

All release channels and locales.

_8) If this data collection is default on, what is the opt-out mechanism for users?_

Users can opt out of of data collection by disabling Usage and technical data from _Settings → Send Usage Data_.

_9) Please provide a general description of how you will analyze this data._

Glean 

_10) Where do you intend to share the results of your analysis?_

Only on Glean, mobile teams and BD have internal access.

_12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?_

No.

File: Client/metrics.yaml [hunk 1]
```diff
@@ -1834,6 +1820,74 @@ widget:
...
+    extra_keys:
+      changed_to:
+        type: string
+        description: |
+          The value the switch was set to.
+    bugs:
+      - https://mozilla-hub.atlassian.net/browse/FXIOS-3487
+    data_reviews:
+      - https://mozilla-hub.atlassian.net/browse/FXIOS-3487
+      - https://github.com/mozilla-mobile/firefox-ios/pull/9673
```
Comment (dnarcese) [inline Client/metrics.yaml:1840]: Lets update this with this PR link

**Comment by dnarcese** [comment]
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes, Firefox iOS provides data controls in the app settings

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This should be re-reviewed w/ Ashley Thomas every 6mo

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 2

_5) Is the data collection request for default-on or default-off?_

Default on

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

File: Client/metrics.yaml (hunk 1 above)
Comment (electricRGB) [reply Client/metrics.yaml:1840]: Updating it in the commit where I'm addressing some of Laurie's things. :)

File: Client/Frontend/Home/Wallpapers/WallpaperManager.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    var wallpapers: [Wallpaper] {
+        return dataManager.availableWallpapers
+    }
+
+    var currentWallpaperImage: UIImage? {
+        return storageManager.retrieveCurrentWallpaperImage()
+    }
+
+    var savedWallpaper: Wallpaper {
```
Comment (lmarceau) [inline Client/Frontend/Home/Wallpapers/WallpaperManager.swift:24]: name

File: Client/Frontend/Home/Wallpapers/WallpaperStorageManager.swift [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    // MARK: - Storage
+    public func store(_ wallpaper: Wallpaper) {
+        store(imageSet: wallpaper.image) { result in
+            switch result {
+            case .success(()):
+                self.store(wallpaperObject: wallpaper)
+                NotificationCenter.default.post(name: .WallpaperDidChange, object: nil)
+            case .failure(let error):
+                print("There was an error storing the wallpaper: ", error.localizedDescription)
```
Comment (lmarceau) [inline Client/Frontend/Home/Wallpapers/WallpaperStorageManager.swift:26]: log

File: Client/Frontend/Settings/HomepageSettings/WallpaperSettings/WallpaperSettingsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import Shared
+import UIKit
+
+private struct WallpaperSettingsUX {
+    static let collectionTitleFontMaxSize = 40.0
+    static let switchTitleFontMaxSize = 46.0
+    static let switchDescriptionFontMaxSize = 34.0
```
Comment (lmarceau) [inline Client/Frontend/Settings/HomepageSettings/WallpaperSettings/WallpaperSettingsViewController.swift:12]: Max font size

**Review by lmarceau — APPROVED** [review]

**Pull request merged**